File: app/src/domain/PuyoPair.ts
import { Puyo } from './Puyo'

export interface Position {
  x: number
  y: number
}

export class PuyoPair {
  public rotation = 0

  constructor(
    public main: Puyo,
    public sub: Puyo,
    public x: number,
    public y: number
  ) {}

  rotate(): void {
    this.rotation = (this.rotation + 90) % 360
  }

  getMainPosition(): Position {
    return { x: this.x, y: this.y }
  }

  getSubPosition(): Position {
    switch (this.rotation) {
      case 0: // 上
        return { x: this.x, y: this.y - 1 }
      case 90: // 右
        return { x: this.x + 1, y: this.y }
      case 180: // 下
        return { x: this.x, y: this.y + 1 }
      case 270: // 左
        return { x: this.x - 1, y: this.y }
      default:
        return { x: this.x, y: this.y - 1 }
    }
  }
}

File: app/fsharp-2/wpf-2/src/PuyoPuyo.WPF/Domain/PuyoPair.fs
module Domain.PuyoPair

open System
open Domain.Puyo

// 位置
type Position = { X: int; Y: int }

// ぷよペア
type PuyoPair =
    { Axis: PuyoColor
      Child: PuyoColor
      AxisPosition: Position
      ChildPosition: Position
      Rotation: int }

// ぷよペアを生成
let generatePuyoPair (random: Random) =
    { Axis = randomPuyoColor random
      Child = randomPuyoColor random
      AxisPosition = { X = 2; Y = 0 }
      ChildPosition = { X = 2; Y = -1 }
      Rotation = 0 }

// ぷよペアの位置と色を取得
let getCurrentPairPuyos (puyoPair: PuyoPair option) =
    match puyoPair with
    | None -> []
    | Some pair ->
        [ (pair.AxisPosition.X, pair.AxisPosition.Y, pair.Axis)
          (pair.ChildPosition.X, pair.ChildPosition.Y, pair.Child) ]
